File: backend/app/middleware/rate_limiter.py
"""
rate_limiter.py — slowapi rate limiting.
Replaces: backend/src/middleware/rateLimiter.ts (express-rate-limit)
"""

from slowapi import Limiter
from slowapi.util import get_remote_address

limiter = Limiter(key_func=get_remote_address)

# Rate limit strings for use as decorators on routes:
# - general:    100/15min  (handled globally in main.py)
# - retrieval:  20/min
# - password:   5/10min
# - upload:     10/hour
RETRIEVAL_LIMIT = "20/minute"
PASSWORD_LIMIT = "5/10minutes"
UPLOAD_LIMIT = "10/hour"

File: backend/src/middleware/rateLimiter.ts
import rateLimit from 'express-rate-limit';

/**
 * Rate limiters for different endpoint categories.
 * Uses in-memory store (swap to Redis store for multi-instance deployments).
 */

// General API rate limit
export const generalLimiter = rateLimit({
  windowMs: 15 * 60 * 1000, // 15 minutes
  max: 100,
  standardHeaders: true,
  legacyHeaders: false,
  message: {
    success: false,
    error: 'Too many requests, please try again later.',
  },
});

// UID retrieval — 20 requests per IP per minute
export const retrievalLimiter = rateLimit({
  windowMs: 60 * 1000, // 1 minute
  max: 20,
  standardHeaders: true,
  legacyHeaders: false,
  message: {
    success: false,
    error: 'Too many retrieval attempts. Please wait a moment.',
  },
});

// Password attempts — 5 per UID per 10 minutes (brute-force protection)
export const passwordLimiter = rateLimit({
  windowMs: 10 * 60 * 1000, // 10 minutes
  max: 5,
  standardHeaders: true,
  legacyHeaders: false,
  keyGenerator: (req) => {
    // Rate limit by IP + UID combination
    const uid = req.params.uid || 'unknown';
    return `${req.ip}:${uid}`;
  },
  message: {
    success: false,
    error: 'Too many password attempts. Please try again in 10 minutes.',
  },
});

// Upload — 10 requests per IP per hour
export const uploadLimiter = rateLimit({
  windowMs: 60 * 60 * 1000, // 1 hour
  max: 10,
  standardHeaders: true,
  legacyHeaders: false,
  message: {
    success: false,
    error: 'Upload limit reached. Please try again later.',
  },
});
